zerver/webhooks/newrelic/view.py

```diff
@@ -1,169 +1,264 @@
+            body_custom_field_detail += invalid_field_message
+    return body_custom_field_detail
+
+
+@webhook_view("NewRelic", all_event_types=INCIDENT_STATES)
 @typed_endpoint
 def api_newrelic_webhook(
     request: HttpRequest,
     user_profile: UserProfile,
     *,
     payload: JsonBodyPayload[WildValue],
 ) -> HttpResponse:
-    # Handle old format
-    # Once old is EOLed, delete if block and keep else block
-    if not payload.get("id").tame(check_none_or(check_string_fixed_length(36))):
-        info = {
-            "condition_name": payload.get("condition_name", "Unknown condition").tame(check_string),
-            "details": payload.get("details", "No details.").tame(check_string),
-            "incident_url": payload.get("incident_url", "https://alerts.newrelic.com").tame(
-                check_string
-            ),
-            "incident_acknowledge_url": payload.get(
-                "incident_acknowledge_url", "https://alerts.newrelic.com"
-            ).tame(check_string),
-            "status": payload.get("current_state", "None").tame(check_string),
-            "iso_timestamp": "",
-            "owner": payload.get("owner", "").tame(check_string),
-        }
-
-        unix_time = payload.get("timestamp").tame(
-            check_none_or(check_union([check_string, check_int]))
-        )
-        if unix_time is None:
-            raise JsonableError(_("The newrelic webhook requires timestamp in milliseconds"))
-
-        info["iso_timestamp"] = str(unix_milliseconds_to_timestamp(unix_time, "newrelic"))
-
-        # Add formatting to the owner field if owner is present
-        if info["owner"] != "":
-            info["owner"] = "by **{}** ".format(info["owner"])
-
-        # These are the three promised current_state values
-        if info["status"].lower() == "open":
-            content = OPEN_TEMPLATE.format(**info)
-        elif info["status"].lower() == "acknowledged":
-            content = DEFAULT_TEMPLATE.format(**info)
-        elif info["status"].lower() == "closed":
-            content = DEFAULT_TEMPLATE.format(**info)
-        else:
-            raise JsonableError(
-                _("The newrelic webhook requires current_state be in [open|acknowledged|closed]")
-            )
+    is_complete = check_essential_fields(payload)
+    if is_complete:
+        message_context = parse_payload(payload)
+        body_detail = NOTIFICATION_BODY_TEMPLATE
 
-        topic_info = {
-            "policy_name": payload.get("policy_name", "Unknown Policy").tame(check_string),
-            "incident_id": payload.get("incident_id", "Unknown ID").tame(
-                check_union([check_string, check_int])
-            ),
-        }
-        topic_name = TOPIC_TEMPLATE.format(**topic_info)
+        additional_info_block = generate_additional_info_message(payload)
```

Comment:
This is just parsing one optional field in the payload. It seems unnecessary to have a separate function for it. I think we could just do it in the main view function, something like ...

```python
        acknowledged_by = ""
        owner = payload.get("owner").tame(check_none_or(check_string))
        if owner is not None:
            acknowledged_by = f"|:silhouette:  Acknowledged by **{owner}**|\n"
```

Reply:
hmm, yeah I was thinking that that function could house other optional fields we might add later But we can move it to the main function for now.  I simplified it a bit, hopefully, it doesn't take away too much readability.

```
owner = payload.get("owner").tame(check_none_or(check_string))
body_detail += f"|:silhouette: Acknowledged by **{owner}**|\n" if owner else ""
```